File: js/set-1/set-1.js
const crypto = require('node:crypto');
const fs = require('fs');
const path = require('path');
const { log } = require('node:console');

function hexToBase64(hexStr) {
    // Turn string into binary from hex and then turn into base64 string
    return Buffer.from(hexStr, 'hex').toString('base64');
}

// XOR each byte to create new array of bytes
function xor(bufferOne, bufferTwo) {
    const length = Math.max(bufferOne.length, bufferTwo.length);
    const buffer = Buffer.alloc(length); // creates new buffer of size of longest input

    for (let i = 0; i < length; i++) {
        buffer[i] = bufferOne[i] ^ bufferTwo[i];
    }

    return buffer;
}

function xorHexStrings(source, comparator) {
    const sourceBytes = Buffer.from(source, 'hex');
    const comparatorBytes = Buffer.from(comparator, 'hex');

    return xor(sourceBytes, comparatorBytes).toString('hex');
}

function isAlphabetChar(byte) {
    return (byte >= 65 && byte <= 90) || (byte >= 97 && byte <= 122) || byte === 92;
}

function isSpaceChar(byte) {
    return byte === 32;
}

function isSpecialChar(byte) {
    return (byte >= 33 && byte <= 64) || (byte >= 93 && byte <= 96) || (byte >= 123 && byte <= 127) || byte === 91;
}

function scoreBytes(buffer) {
    let score = 0;

    for (const byte of buffer) {
        if (isAlphabetChar(byte)) {
            score += 20;
        } else if (isSpaceChar(byte)) {
            score += 100;
        } else if (isSpecialChar(byte)) {
            score += 10;
        }
        else {
            score -= 100;
        }
    }

    return score;
}

/**
 * Determines which single-byte key was used to encrypt the given hex string.
 *
 * @param {string} hexString - The hexadecimal string to analyze.
 * @returns {number} The encryption key used as a decimal value.
 */
function findEncryptionKey(hexString) {
    const largestHexLiteral = 0xFF; // 255
    let encryptionKey = '';
    let hiddenMessage = '';
    let highestScore = 0;

    /* For each byte in all possible hex bytes, XOR against the input "hexString"'s bytes
    to produce a decrypted message */
    for (let key = 0; key <= largestHexLiteral; key++) {
        const { score, decryptedText } = _decryptHexLineWithKey(hexString, key);

        if (score > highestScore) {
            highestScore = score;
            encryptionKey = key;
            hiddenMessage = decryptedText;
        }
    }

    return {
        key: encryptionKey,
        score: highestScore,
        text: hiddenMessage
    };
}

/**
 * Single Byte XOR - Decrypts each byte of hexString with the given key and scores the decrypted text against English text
 *
 * @param {string} hexString - The hexadecimal string to analyze.
 * @param {number} key - The key to decrypt each byte of hexString against
 * @returns {{score: number, key: number, decryptedText: string}} An object containing the score, key, and decrypted text.
 */
function _decryptHexLineWithKey(hexString, key) {
    const decryptedBytes = Buffer.from(hexString, 'hex').map((byte) => byte ^ key); // decrypt each byte
    const decryptedText = Buffer.from(decryptedBytes).toString('utf8'); // turn each byte into utf8 (text)
    const score = scoreBytes(decryptedBytes);

    return {
        score,
        key,
        decryptedText,
    }
}

/**
 * Decrypts a hexadecimal string using a single-byte XOR key.
 *
 * @param {string} hexString - The hexadecimal string to decrypt.
 * @param {number} key - The single-byte XOR key used for decryption.
 * @returns {string} The decrypted string.
 */
function xorDecrypt(hexString, key) {
    const hex = Buffer.from(hexString, 'hex');

    let result = '';

    for (let i = 0; i < hex.length; i++) {
        result += String.fromCharCode(hex[i] ^ key);
    }

    return result;
}


/**
 * Reads file, for each hex line decrypts with all possible hex bytes. *
 * Returns the key (byte) that finds the decrypted text most similar to english text
 *
 * @param {string} fileName - The name of the file to read.
 * @returns {number} The byte that was used to encrypt (XORing) in decimal.
 */
async function findEncyptionKeyInFile(fileName) {
    const file = fs.readFileSync(path.join(__dirname, '..', 'data', fileName), 'utf-8');
    const lines = file.split('\n');

    let highestScore = 0;
    let foundKey = null;
    let englishText = '';

    for (const line of lines) {
        const { key, score, text } = findEncryptionKey(line);

        if (score > highestScore) {
            highestScore = score;
            foundKey = key;
            englishText = text;
        }
    }

    console.log('Found text: ', englishText);
    console.log('highestScore: ', highestScore);
    return foundKey;
};


/**
 * Reads file, for each hex line decrypts with the provided key. *
 * Returns the decrypted line that mostly resembles english text from the provided file. 
 *
 * @param {string} fileName - The name of the file to read.
 * @returns {number} The byte that was used to encrypt (XORing) in decimal.
 */
function findTextFromFileWithKey(fileName, key) {
    const file = fs.readFileSync(path.join(__dirname, '..', 'data', fileName), 'utf-8');
    const lines = file.split('\n');

    let foundText = '';
    let highestScore = 0;

    for (const line of lines) {
        const { score, decryptedText } = _decryptHexLineWithKey(line, key);

        if (score > highestScore) {
            highestScore = score;
            foundText = decryptedText;
        }
    }

    return foundText;
}

/**
 * Sequentially XOR each byte of the key to the plainText. *
 * Returns the encrypted string result of the sequential XOR.
 *
 * @param {string} plainText - The value to encrypt
 * @param {string} key - The value to encrypt against
 * @returns {number} The result of XOR'ing 
 */
function repeatingKeyXOR(plainText, key) {
    const plainTextBytes = Buffer.from(plainText, 'utf8');
    const keyBytes = Buffer.from(key, 'utf8');
    const buffer = Buffer.alloc(plainTextBytes.length);

    // index:    0, 1, 2, 3, 4, 5, 6, 7, 8
    // keyIndex: 0, 1, 2, 0, 1, 2, 0, 1, 2

    for (let i = 0; i < plainTextBytes.length; i++) {
        const keyIndex = i % key.length;

       // xor them 
        buffer[i] = plainTextBytes[i] ^ keyBytes[keyIndex];
    }

    return buffer.toString('hex');
}

function getHammingDistance(aBytes, bBytes) {
    const length = Math.max(aBytes.length, bBytes.length);

    let differingBitCount = 0;

    for (let i = 0; i < length; i++) {
        // how this works. take these 2 nibbles:
        // 1:   0101 
        // 2:   1101
        // xor: 1000

        // xor & 1: compares each rightmost digit with 1. If 1 if both are 1

        // 1000 >>= 1 shifts bits 1 to the right:
        // ex:
        //   0100 
        //   0010
        //   0001
        //   0000

        // look at each byte:
        const byteA = aBytes[i] || 0;
        const byteB = bBytes[i] || 0;

        let xor = byteA ^ byteB; // get the diff (xor)

        while (xor !== 0) { // all shifted to the right (0000)
            differingBitCount += xor & 1; // is the right most digit a 1
            xor >>= 1; // shift 1 to the right to look at next bits
        }
    }

    return differingBitCount;
}

// Takes in binary data and split it into incremental chunks.
// Finds the chunk size with the lowest average hamming distance (most similar bits)
function _findProbableKeySize(cipherData) {
    let keySizeWithSmallestHammingDistance = 2;
    let smallestNormalizedDistance = Infinity;

    // create 4 chunks of size key and get hemming distance against each other
    // keySize that has the shortest average hemming distance is the found keySize
    for (let keySize = 2; keySize <= 40; keySize++) {
        const chunks = [
            cipherData.subarray(0, keySize),
            cipherData.subarray(keySize, keySize * 2),
            cipherData.subarray(keySize * 2, keySize * 3),
            cipherData.subarray(keySize * 3, keySize * 4),
        ];
        const averagesForKey = [];

        for (let i = 0; i < chunks.length; i++) {
            const chunk = chunks[i];

            // look at other chunks and get hamming distance of them
            for (let j = 0; j < chunks.length; j++) {
                if (j === i) {
                    continue;
                }

                const otherChunk = chunks[j];

                const hammingDistance = getHammingDistance(chunk, otherChunk);
                const normalizedDistance = hammingDistance / keySize;

                averagesForKey.push(normalizedDistance);
            }
        }

        const average = averagesForKey.reduce((sum, curr) => sum + curr, 0) / averagesForKey.length;

        if (average < smallestNormalizedDistance) {
            keySizeWithSmallestHammingDistance = keySize;
            smallestNormalizedDistance = average;
        }

    }

    return keySizeWithSmallestHammingDistance;
}

function _readAndBase64Decode(fileName) {
    const file = fs.readFileSync(path.join(__dirname, '..', 'data', fileName), 'utf-8');

    return Buffer.from(file, 'base64');
}

// Make a block, of size keysize, that is the first byte of every block, and a block that is the second byte of every block, and so on.
function _transposeKeySizedBlocks(data, keySize) {
    const blocks = Array.from({ length: keySize }, () => []);

    for (let i = 0; i < data.length; i++) {
        blocks[i % keySize].push(data[i]);
    }

    return blocks.map(block => Buffer.from(block));
}

// Turn into hex, find encryption key, build up string from bytes (keys)
function _determineKey(transposedBlocks) {
    let key = '';

    for (const block of transposedBlocks) {
        const hexBlock = block.toString('hex');
        const { key: singleByteKey } = findEncryptionKey(hexBlock);
        
        key += String.fromCharCode(singleByteKey);
    }

    return key;
}

// Finds the key used to encrypt the file
function breakRepeatingKeyXOR(fileName) {
    const cipherData = _readAndBase64Decode(fileName);
    const keySize = _findProbableKeySize(cipherData);
    const transposedBlocks = _transposeKeySizedBlocks(cipherData, keySize);

    return _determineKey(transposedBlocks);
}

function repeatingKeyXORForFile(fileName, key) {
    const cipherData = _readAndBase64Decode(fileName);

    // This is doing what repeatingKeyXOR is doing:
    const keyBytes = Buffer.from(key, 'utf8');
    const buffer = Buffer.alloc(cipherData.length);

    for (let i = 0; i < cipherData.length; i++) {
        const keyIndex = i % key.length;

       // xor them 
        buffer[i] = cipherData[i] ^ keyBytes[keyIndex];
    }

    return buffer.toString('utf8');
}

// AES decrypt
// AES: symmetric block cipher that decrypts in fixed-sized blocks (e.g., 128 bits for AES)
// ECB mode: encrypts each block independently
// symmetric block cipher: type of cipher that uses same key for encryption/decryption
    // block cipher: processed in fixed sized chunks
function decryptAES(cipherText, key) {
    const decipher = crypto.createDecipheriv('aes-128-ecb', key, null);

    let result = decipher.update(cipherText, 'binary', 'utf8');

    result += decipher.final('utf8');

    return result;
}

function decryptFileAESinECBmode(fileName, key) {
    // base64 decode the ciphertext
    const cipherText = _readAndBase64Decode(fileName);
    
    // AES decrypt
    return decryptAES(cipherText, key);
}

/**
 * Reads a file with hex-encoded AES-ECB encrypted ciphertexts, 
 * converts each line into a byte array, and splits each byte array into blocks of 16 bytes.
 *
 * @param {string} fileName - The name of the file to be read. Each line in the file is a hex-encoded ciphertext.
 * @returns {Array<Array<Buffer>>} - A list of lists where each inner array represents a line of ciphertext 
 *                                   divided into 16-byte blocks.
 *                                   Format: [[[16Bytes, 16Bytes, ...]], [[16Bytes, 16Bytes, ...]], ...]
 */
function _readFileAndTransposeIntoBlocksOf16Bytes(fileName) {
    const file = fs.readFileSync(path.join(__dirname, '..', 'data', fileName), 'utf-8'); // simply read in the file as text
    const buffers = file.split('\n').map(line => Buffer.from(line, 'hex')); // create a list of buffers that hold bytes for each line (they are hex, so treat them accordingly)

    // Create the final structure: a list of lists where each inner list contains 16-byte blocks
    const buffersInBlocksOf16Bytes = buffers.map(buffer => {
        const blocks = [];

        for (let i = 0; i < buffer.length; i += 16) {
            blocks.push(buffer.subarray(i, i + 16)); // Split buffer into 16-byte blocks
        }

        return blocks;
    });

    return buffersInBlocksOf16Bytes;
}

// check if block shows up in blocks more than once
// blocks: [[16Bytes], [16Bytes], ...]]
// blocks represents a line. An array of Buffers
// if there is a buffer that shows up more than once, returns true
function checkContainsDuplicates(buffers) {
    const visited = [];

    for (let i = 0; i < buffers.length; i++) {
        const buffer = buffers[i];

        if (visited.some((visitedBuffer) => visitedBuffer.equals(buffer))) {
            return true;
        }

        visited.push(buffer);
    }

    return false;
}

/*
1. Convert Hex-Encoded Ciphertext into Bytes:
    The ciphertexts provided are in hexadecimal form. You’ll need to convert each hex string into its corresponding byte array for processing.

2. Divide Each Ciphertext into Blocks:
    AES uses a block size of 16 bytes (128 bits). After converting the ciphertext to bytes, you need to divide it into 16-byte blocks.
    
3. Detect Repeated Blocks:
    ECB mode will produce identical ciphertext blocks for identical plaintext blocks. This is a key characteristic of ECB’s deterministic nature.
    Look for repeated blocks within each ciphertext. If a ciphertext has any repeated 16-byte blocks, it’s very likely encrypted with ECB.

4. Identify the Ciphertext with Repeats:
    The ciphertext that contains repeated blocks is the one encrypted with ECB. Return or print the index or the actual ciphertext.
*/
function detectAESinECB(fileName) {
    const ciphersInBlocksOf16Bytes = _readFileAndTransposeIntoBlocksOf16Bytes(fileName);

    for (let i = 0; i < ciphersInBlocksOf16Bytes.length; i++) {
        const line = ciphersInBlocksOf16Bytes[i];

        // check if this line has any duplicates, if so we found the one encrypted with AES in ECB mode
        if (checkContainsDuplicates(line)) {
            return { index: i, line }
        }
    }

    return { index: -1, line: '' };
}

module.exports = {
    hexToBase64,
    xorHexStrings,
    findEncryptionKey,
    xorDecrypt,
    findEncyptionKeyInFile,
    findTextFromFileWithKey,
    repeatingKeyXOR,
    getHammingDistance,
    breakRepeatingKeyXOR,
    repeatingKeyXORForFile,
    decryptFileAESinECBmode,
    detectAESinECB,
};

// 1111
// binary: 16 8 4 2 1
//                 2^2(4) | 2^1(2) | 2^0(1)

// Binary to Hex -----------------------
//              0001 0011 0101 
// decimal:        1    3    5
// hex:            1    3    5
// result:                 135

// 1001 0111 1101
//    9    7   13
//    9    7    D

File: go/set-1/set-1.go
package main

import (
	"bufio"
	"crypto/aes"
	"encoding/base64"
	"encoding/hex"
	// "fmt"
	"log"
	"math"
	"os"
	"path/filepath"
)

func HexToBase64(input string) string {
	// turn hex to array of bytes
	bytes, err := hex.DecodeString(input)

	if err != nil {
		log.Fatal(err)
	}

	// turn array of bytes into base64
	return base64.StdEncoding.EncodeToString(bytes)
}

func XORHexStrings(source string, comparator string) string {
	sourceBytes, err := hex.DecodeString(source)

	if err != nil {
		log.Fatal(err)
	}

	comparatorBytes, err := hex.DecodeString(comparator)

	if err != nil {
		log.Fatal(err)
	}

	var maxLength = int(math.Max(float64(len(sourceBytes)), float64(len(comparatorBytes))))
	var xordBytes []byte = make([]byte, maxLength)

	for i := 0; i < maxLength; i++ {
		xordBytes[i] = sourceBytes[i] ^ comparatorBytes[i]
	}

	return hex.EncodeToString(xordBytes)
}

func getDecryptedBytes(hexInput string, key int) []byte {
	inputBytes, err := hex.DecodeString(hexInput)

	if err != nil {
		log.Fatal(err)
	}

	var decryptedBytes []byte = make([]byte, 0, len(inputBytes))

	for _, inputByte := range inputBytes {
		decrypted := inputByte ^ byte(key)
		decryptedBytes = append(decryptedBytes, byte(decrypted))
	}

	return decryptedBytes
}

func GetKeyAndScoreForLine(hexInput string) (int, int) {
	const largestHex = 0xFF

	var topScore int = 0
	var foundEncryptionKey = 0

	for key := 0; key < largestHex; key++ {
		var decryptedBytes []byte = getDecryptedBytes(hexInput, key)

		score := scoreBytes(decryptedBytes)

		if score > topScore {
			topScore = score
			foundEncryptionKey = key
		}
	}

	return foundEncryptionKey, topScore
}

func isAlphabetChar(bite byte) bool {
	return (bite >= 64 && bite <= 90) || (bite >= 97 && bite <= 122) || bite == 92
}

func isSpaceChar(bite byte) bool {
	return bite == 32
}

func isSpecialChar(bite byte) bool {
	return (bite >= 33 && bite <= 64) || (bite >= 93 && bite <= 96) || (bite >= 123 && bite <= 127) || bite == 91
}

func scoreBytes(buffer []byte) int {
	score := 0

	for _, bite := range buffer {
		bitten := byte(bite)

		switch {
		case isAlphabetChar(bitten):
			score += 20
		case isSpaceChar(bitten):
			score += 100
		case isSpecialChar(bitten):
			score += 10
		default:
			score -= 100
		}
	}

	return score
}

func FindEncyptionKeyInFile(fileName string) int {
	file, err := os.Open(filepath.Join("..", "data", fileName))

	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	topScore := 0
	bestKey := 0

	for scanner.Scan() {
		line := scanner.Text()
		key, score := GetKeyAndScoreForLine(line)

		if score > topScore {
			topScore = score
			bestKey = key
		}
	}

	if err := scanner.Err(); err != nil {
		log.Println("Error reading file:", err)
	}

	return bestKey
}

func FindTextFromFileWithKey(fileName string, key int) string {
	file, err := os.Open(filepath.Join("..", "data", fileName))

	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	topScore := 0
	bestText := ""

	for scanner.Scan() {
		line := scanner.Text()
		decryptedBytes := getDecryptedBytes(line, key)
		text := string(decryptedBytes)

		score := scoreBytes(decryptedBytes)
		if score > topScore {
			topScore = score
			bestText = text
		}
	}

	if err := scanner.Err(); err != nil {
		log.Println("Error reading file:", err)
	}

	return bestText
}

/*
 	* Sequentially XOR each byte of the key to the plainText. *
		* Ex: Text: Hello; Key: ICE
			* H ^ I, e ^ C, l ^ E, l ^ I... etc
 	* Returns the encrypted string result of the sequential XOR.
*/
func RepeatingKeyXOR(text string, key string) string {
	textBytes := []byte(text)
	keyBytes := []byte(key)

	encryptedBytes := make([]byte, 0, len(textBytes))

	for i, bite := range textBytes {
		keyIndex := i % len(keyBytes)
		encryptedBytes = append(encryptedBytes, bite ^ keyBytes[keyIndex])
	}

	return hex.EncodeToString(encryptedBytes)
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

// Gets amount of differing bites for aBytes and bBytes
func GetHammingDistance(aBytes []byte, bBytes []byte) int {
	length := max(len(aBytes), len(bBytes))

	differingBitCount := 0

	for i := 0; i < length; i++ {
		xor := aBytes[i] ^ bBytes[i]

		for xor != 0 {
			differingBitCount += int(xor & 1)
			xor = xor >> 1
		}
	}

	return differingBitCount
}

func readFileAsBytes(fileName string) []byte {
	file, err := os.Open(filepath.Join("..", "data", fileName))

	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	var data []byte

	for scanner.Scan() {
		data = append(data, scanner.Bytes()...)
	}

	if err := scanner.Err(); err != nil {
		log.Println("Error reading file:", err)
	}

	return data
}

/*
	- Tries to discover key length
	- Breaks the data into chunks of estimated keysize (2 - 40)
	- compares hamming distance of 2 consecutive chunks of size "keysize" and gets an average hamming distance of that keysize
	- key length with lowest hamming distance is probably the key
*/
func findProbableKeyLength(data []byte) int {
	smallestAverage := math.MaxFloat64
	bestKey := 2

	for maybeKeySize := 2; maybeKeySize <= 41; maybeKeySize++ { // for each potential keySize
		amtOfChunks := len(data) / maybeKeySize
		averagesForKey := make([]float64, 0)

		// build up distances per key
		for i := 0; i < amtOfChunks - 1; i++ {
			chunkOne := data[i * maybeKeySize : (i + 1) * maybeKeySize]
			chunkTwo := data[(i + 1) * maybeKeySize : (i + 2) * maybeKeySize]

			distance := GetHammingDistance(chunkOne, chunkTwo)
			aveDistancePerKey := float64(distance) / float64(maybeKeySize)
			averagesForKey = append(averagesForKey, aveDistancePerKey)
		} 
		
		// determine best key (one with the smallest total average)
		sum := float64(0)
		for _, ave := range averagesForKey {
			sum += ave
		}
		aveForKey := sum / float64(len(averagesForKey))

		if (aveForKey < smallestAverage) {
			bestKey = maybeKeySize
			smallestAverage = aveForKey
		}
	}

	return bestKey
}

/*
	- Breaks bytes into blocks that were encrypted using same single byte xor
	- returns new blocks where 1st block is the 1st byte of every block, 2nd is 2nd byte of every keySize block, and so on
	Ex:
	data: 			[abc123defg456]
	keysize blocks: [abc, 123, def, g45, 6]
	transposed: 	[a1dg6, b2d4, c3f5] each block here was encrypted with same key
		Example: 
		- Imagine the key was "KEY"
		- block at index 0 was encrtyped with char "K"
		- block at index 1 with "E"
		- block at index 2 with "Y"
*/
func TransposeBlocks(data []byte, keySize int) [][]byte {
	blocksEncryptedBySameKey := make([][]byte, keySize) // will have "keySize" amt of elements

	for i := 0; i < len(data); i++ {
		// append the byte into the block at "i % keySize"
		blocksEncryptedBySameKey[i % keySize] = append(blocksEncryptedBySameKey[i % keySize], data[i])
	}

	return blocksEncryptedBySameKey
}

/*
	- Solve each block as if it were single-char-xor.
	- Turn each block into hex and run it through GetKeyAndScoreForLine
	- Builds up each key as a string and returns it
*/
func getKeyFromBlocks(transposedBlocks [][]byte) string {

	keyBytes := make([]byte, 0)	

	for _, block := range transposedBlocks {
		key, _ := GetKeyAndScoreForLine(hex.EncodeToString(block))
		keyBytes = append(keyBytes, byte(key))
	}	

	return string(keyBytes)
}

func decodeBase64(data []byte) []byte {
	decoded, err := base64.StdEncoding.DecodeString(string(data))

	if err != nil {
		log.Fatalf("Base64 decoding error: %v", err)
	}

	return decoded
}

/* 
	- Reads a file that has been repeating key XOR encrypted and then base64 encoded.
	- Discovers the key used to encrypt the file
*/ 
func BreakRepeatingKeyXOR(fileName string) string {
	// Read the file, turns it into bytes, then decode it from bas64
	cipherData := readFileAsBytes(fileName)
	decodedCipherData := decodeBase64(cipherData)

	// Find the probable key length
	keySize := findProbableKeyLength(decodedCipherData)

	blocksEncryptedBySameKey := TransposeBlocks(decodedCipherData, keySize)

	return getKeyFromBlocks(blocksEncryptedBySameKey)
}

func DecryptAES(data []byte, key string) string {
	cipher, err := aes.NewCipher([]byte(key))
	if err != nil {
		log.Fatal(err)
	}

	blockSize := len(key) // 16
	plainText := make([]byte, len(data))
	amtOfBlocks := len(plainText) / blockSize

	// break data into key-sized chunks and decrypt them chunk by chunk (ECB mode)
	for i := 0; i < amtOfBlocks; i++ {
		start := i * blockSize // 0
		end := (i + 1) * blockSize // 16

		cipher.Decrypt(plainText[start:end], data[start:end])
	}

	return string(plainText)
}

func DecryptFileAESinECBmode(fileName string, key string) string {
	// Read file and decode base64
	data := readFileAsBytes(fileName)
	decoded := decodeBase64(data)

	// Decrypt AES
	return DecryptAES(decoded, key)
}

func checkLineForDuplicates(blocks [][]byte) bool {
    // Use a map to track seen blocks
    seenBlocks := make(map[string]bool)

    // Iterate over the blocks
    for _, block := range blocks {
        // Convert the block to a string to use as a map key
        blockStr := string(block)

        // Check if the block has been seen before
        if _, exists := seenBlocks[blockStr]; exists {
            return true // Duplicate found
        }

        // Mark the block as seen
        seenBlocks[blockStr] = true
    }

    return false // No duplicates found
}

type AesECBDetection struct {
	index int
	line []byte
}

func findLineWithDuplicateBlocks(transposedLines [][][]byte, lines [][]byte) AesECBDetection {
	for i, line := range transposedLines {
		if hasDuplicates := checkLineForDuplicates(line); hasDuplicates {
			return AesECBDetection{
				index: i,
				line: lines[i],
			}
		}
	}

	return AesECBDetection{}
}

/* 
	- Reads file as independent lines
	- Turns each line into a list of lists of 16bytes:
		[
			[[16bytes], [16bytes], [16bytes]] line
			[[16bytes], [16bytes], [16bytes]] line
			...
		]
	- Figures out which line has duplicates and returns the index of that line and the line itself
*/
func DetectAESinECB(fileName string) AesECBDetection {
	file, err := os.Open(filepath.Join("..", "data", fileName))

	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	lines := make([][]byte, 0)
	transposedLines := make([][][]byte, 0)

	for scanner.Scan() {
		line := scanner.Bytes()

		// make copy of line and append into lines (scanner reuses internal buffer. Could create duplicates)
		lineCopy := make([]byte, len(line))
		copy(lineCopy, line)

		// populate the initial lines to return the line at the end with the index
		lines = append(lines, lineCopy)

		// turn lineCopy into [][]byte with each inner []byte containing 16 bytes
		lineInChunksOf16Bytes := make([][]byte, 0)

		for start := 0; start < len(lineCopy); start += 16 {
			end := start + 16

			// if out of bounds, set end to the last index + 1 (non inclusive end)
			if end > len(lineCopy) {
				end = len(lineCopy)
			}

			chunk := lineCopy[start:end]

			lineInChunksOf16Bytes = append(lineInChunksOf16Bytes, chunk)
		}

		transposedLines = append(transposedLines, lineInChunksOf16Bytes)
	}

	if err := scanner.Err(); err != nil {
		log.Println("Error reading file:", err)
	}

	return findLineWithDuplicateBlocks(transposedLines, lines)
}